---
title: "wind: wORKFLOW FOR PiRNAs AnD BEYONd"
subtitle: "Computational workflow for Data Exploration resulted from smallRNA-seq of testis and COLO205 samples"
author: "Constantinos Yeles (Konstantinos Geles)"
date: "`r format(Sys.time(), '%a %b %d %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    theme: paper 
  pdf_document:
    toc: yes
    toc_depth: 3
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```
Following the step 2 we are working again on the docker of Rstudio loaded before.
## 1. Load libraries

```{r load libraries}
suppressPackageStartupMessages({
  library('tidyverse') 
  library('data.table')
  library('plyranges')
  library('tximport')
  library('edgeR')
  library('NOISeq')
  library('rafalib')
  library('pheatmap')
  library('RColorBrewer')
  library('jsonlite')
})
```
## 2. Directory generation for the resulted files
### i. Add date

Used as an identifier for the folder 

```{r todate_of_analysis}
todate <- format(Sys.time(), "%d_%b_%Y")
```
### ii. Make the directory for the results of the exploratory data analysis

```{r make dirs}
my_basename <- file.path("Datasets_analysis", "Testis_COLO205") ## INPUT name of the main folder 
my_exp <- "spike_ins_COLO205_Testis" ## INPUT name of the analysis
genome_input <- "GRCh38" ## INPUT genome version here
my_tools <- c("salmon","featureCounts")
dat_path <- file.path(my_basename, str_glue("EDA_{my_exp}_{genome_input}_{todate}"),
                      my_tools)
dat_path %>% map(~dir.create(., recursive = TRUE))
```
## 3. Make or import the targets file.

If you have used the [fastq-dl](https://github.com/rpetit3/fastq-dl) tool to download the samples from the European Nucleotide Archive then in the folder you have downloaded should be also json file(s) that can be imported.
Otherwise you have to make them from scratch.
The targets file has to have at least three columns with the column names: "sample_name","group","batch"

```{r targets file}
## import the json files in a list and collapse them
targets_file <- file.path(my_basename,"downloaded_samples") %>% 
  list.files( pattern = ".json", recursive = TRUE, full.names = TRUE) %>% 
  map(~jsonlite::fromJSON(.)) %>% 
  bind_rows() %>% 
  as_tibble() %>% 
  select(read_count, run_accession, #select only the columns with basic info
         sample_title, study_accession) %>% 
  mutate(group = sample_title %>% str_remove("_[:digit:]$"), 
         group = group %>% str_remove("_[ABC]_NT$"),
         batch = sample_title %>% str_remove(".+(?=[:digit:]$)"),
         read_count = as.integer(read_count),
         across(.cols = where(is.character), as_factor)) %>% 
  dplyr::rename(sample_name = sample_title)

targets_file$group %>% levels()
```
## 4. Import the salmon files

```{r import salmon}
# load salmon files----
files_salm <- list.files(path = my_basename, pattern = ".sf",
  recursive = TRUE, full.names = TRUE)

names(files_salm) <- files_salm %>% 
  str_remove(".quant.sf") %>% 
  basename() %>% 
  str_remove("_quant")

# tximport-------
txi <- tximport::tximport(files_salm, type = "salmon",
  txOut = TRUE, countsFromAbundance = "lengthScaledTPM")
```
## 5. Make a DGElist object for salmon

```{r DGElist salmon}
# DGElist
# from https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html
# we follow the instructions to import for edgeR 
cts <- txi$counts
normMat <- txi$length

# Obtaining per-observation scaling factors for length, adjusted to avoid
# changing the magnitude of the counts
normMat <- normMat/exp(rowMeans(log(normMat)))
normCts <- cts/normMat

# Computing effective library sizes from scaled counts, to account for
# composition biases between samples
eff.lib <- calcNormFactors(normCts) * colSums(normCts)

# Combining effective library sizes with the length factors, and calculating
# offsets for a log-link GLM
normMat <- sweep(normMat, 2, eff.lib, "*")
normMat <- log(normMat)

# Creating a DGEList object for use in edgeR.
dgl_salmon <- DGEList(cts, samples = targets_file) %>% 
  scaleOffset( normMat) %>% 
  write_rds(file.path(dat_path[1],"dgl_edgeR_salmon.rds"))
# remove objects.
rm(cts, normCts, normMat, txi)
```
## 6. Import the featureCounts object and make a DGElist object

```{r DGElist FeatureCounts}
# load the rds from featureCOunts----
# INPUT rds featureCOunts
fc <- list.files(path = my_basename, recursive = TRUE, 
  pattern = ".+counts.+.rds", 
  full.names = TRUE) %>% 
  read_rds()

# write the matrix for the analysis, annotation stats-----
colnames(fc$counts) <- targets_file$run_accession

fc$counts %>% 
  as_tibble(rownames = "sRNA") %>% 
  write_tsv(file.path(dat_path[2],"raw_reads_fc.txt"))

fc$annotation %>% 
  as_tibble() %>% 
  write_tsv(file.path(dat_path[2],"annotation_fc.txt"))

fc$stat %>% 
  as_tibble() %>% 
  write_tsv(file.path(dat_path[2],"stats_fc.txt"))

dgl_fc <- DGEList(counts = fc$counts,
               samples = targets_file,
               lib.size = colSums(fc$counts),
               norm.factors = rep(1,ncol(fc$counts)))

# give colours to samples ----
pal1 <- tibble(value = c('#e41a1c','#377eb8',
                         '#4daf4a','#984ea3',
                         '#ff7f00','#ffff33',
                         '#a65628','#f781bf')) %>%
  dplyr::slice(1:length(levels(as_factor(targets_file$group)))) %>%
  mutate(
    group = as_factor(levels(as_factor(targets_file$group))))

dgl_fc$colours <- as_factor(inner_join(dgl_fc$samples, pal1,by= "group")$value)

# remove objects ----
rm(pal1)
```
























## 3. Import the expr values
### i. 
```{r}
# Find all rows where ANY numeric variable is greater than zero
rowAny <- function(x) rowSums(x) > 1

# load the gtf 
annot_tbl <- read_gff2("../genome_transc_human/ncRNA_transcripts_100bp_RNA_Central_piRNAbank_hg38.gtf") %>% 
  filter(gene_type == "piRNA")
# load the expr per group
piRNA_grouped_exprs <- list.files(path = my_basename,
                                          pattern = "salmon_FC_cpm_union_grouped.txt", 
                                          full.names = TRUE, 
                                          recursive = TRUE) %>% 
  read_tsv() %>% 
  filter(sncRNA %in% annot_tbl$gene_id,
         rowAny(across(where(is.numeric), ~ .x >= 1))) 

# pick only piRNAs with T->1st base and A ->10nth
piRNAs_possible_ping_pong <- annot_tbl %>% 
  select(gene_id, seq_RNA) %>% 
  filter(str_detect(seq_RNA, "^T[TCGA]{8}A.+"))

Testis_10_top_piRNA <- piRNA_grouped_exprs %>% 
  filter(sncRNA %in% piRNAs_possible_ping_pong$gene_id) %>% 
  arrange(desc(Testis_pool_fc)) %>% 
  select(sncRNA, Testis_pool_fc, COLO205_fc) %>% 
  head(10) %>% .$sncRNA

COLO205_10_top_piRNA <- piRNA_grouped_exprs %>%
  filter(sncRNA %in% piRNAs_possible_ping_pong$gene_id) %>%
  arrange(desc(COLO205_fc)) %>% 
  select(sncRNA, Testis_pool_fc, COLO205_fc) %>% 
  head(10) %>% .$sncRNA

```

### ii. check the Genomic Ranges from most expressed piRNAs
```{r}
annot_tbl %>% filter(gene_id == Testis_10_top_piRNA[2])
annot_tbl %>% filter(gene_id == COLO205_10_top_piRNA[2])


wh <- annot_tbl %>% 
  filter(gene_id == COLO205_10_top_piRNA[2]) %>% 
  as_tibble() %>% 
  mutate(strand = "*") %>% 
  as_granges()



```
### iii. make a ggbio plot 
```{r}
library("BSgenome.Hsapiens.UCSC.hg38")
library(ggbio)

bg <- BSgenome.Hsapiens.UCSC.hg38
# import the bam of testis sample
testis_pool_bam <- "../testis_colo_pub_workflow/star/Human_Testis_Pool_Non_Treated_align/Human_Testis_Pool_Non_Treated_Aligned.sortedByCoord.out.bam"
Rsamtools::indexBam(testis_pool_bam)

testis_pool_bam <- "../testis_colo_pub_workflow/star/Human_Testis_Pool_Non_Treated_align/Human_Testis_Pool_Non_Treated_Aligned.sortedByCoord.out.bam" %>%
  read_bam %>% 
  select(seq) %>%
  filter_by_overlaps(wh,!is_unmapped_query)

testis_pool_FC_bam <- "../testis_colo_pub_workflow/star/Human_Testis_Pool_Non_Treated_Aligned.sortedByCoord.out.bam.featureCounts.bam" 

Rsamtools::sortBam(testis_pool_FC_bam, 
        destination ="../testis_colo_pub_workflow/star/Human_Testis_Pool_Non_Treated_Aligned.sortedByCoord.out.bam.featureCounts.sorted.bam")

testis_pool_FC_bam <- "../testis_colo_pub_workflow/star/Human_Testis_Pool_Non_Treated_Aligned.sortedByCoord.out.bam.featureCounts.sorted.bam.bam"
Rsamtools::indexBam(testis_pool_FC_bam)

testis_pool_FC_bam_read <-testis_pool_FC_bam %>%
  read_bam %>% 
  select(XS,XT,seq) %>%
  filter_by_overlaps(wh, minoverlap = 10L)

 wh %>% select(seq_RNA,gene_id) %>% as_tibble
aligned_smallRNAs<-testis_pool_FC_bam_read %>% select(XT, cigar) %>% as_tibble %>% count(XT) # %>% arrange(desc(width),seq)
# the plot
p.mis <- autoplot(testis_pool_bam, 
                  bsgenome = bg, 
                  which = wh, 
                  stat = "mismatch")
```

