---
title: "wind: wORKFLOW FOR PiRNAs AnD BEYONd"
subtitle: "Computational workflow for the preprocessing of selected samples from TCGA regarding Breast Cancer"
author: "Constantinos Yeles (Konstantinos Geles)"
date: "`r format(Sys.time(), '%a %b %d %Y')`"
output:
  pdf_document:
    toc: yes
    toc_depth: 3
  html_document:
    toc: yes
    toc_depth: 3
    theme: paper 
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```
## The Data set

We will work with a dataset from TCGA which is a subset of 18 samples from [TCGA-BRCA](https://doi.org/10.1038/ng.2764). In detail, we used 9 Primary Solid Tumors versus 9 Solid Tissue Normal matched samples.  

## Data aqcuisition and preprocessing  

### i. Downloading the samples  

To obtain the data from TCGA we used a manifest with the selected samples and followed the instructions on [GDC website](https://docs.gdc.cancer.gov/Data_Transfer_Tool/Users_Guide/Preparing_for_Data_Download_and_Upload/)

### ii. BAM to FASTQ

The acquired files were in BAM format, in order to perform the whole workflow we transformed them to FASTQ format.

```{bash bam2fq}
docker run --rm -ti -v $(pwd):/home/my_data  congelos/sncrna_workflow

mkdir my_data/fastq_files
for file in my_data/bam_files/*.bam; 
do regex="${file%%.bam}"; samp=`basename ${regex}`; 
echo "Processing sample ${samp}";
samtools bam2fq -@ 6  $file > my_data/fastq_files/${samp}.fastq
echo "pigz sample ${samp}";
pigz --best "my_data/fastq_files/${samp}.fastq"
done
```
### iii. Preprocessing of the samples
We perform quality control(QC) on the fastq files to get basic information about the samples.
We work with the __[Fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)__ tool to perform QC.

```{bash preprocessing}
ANALYSIS_FOLDER="my_data/Datasets_analysis/TCGA_BRCA"

mkdir "${ANALYSIS_FOLDER}"/qc_before_trim \
"${ANALYSIS_FOLDER}"/quants  "${ANALYSIS_FOLDER}"/star

'fastqc' --threads 9 --outdir=${ANALYSIS_FOLDER}/qc_before_trim "${ANALYSIS_FOLDER}"/downloaded_samples/*fastq.gz 

exit
```
## Alignment and Quantification
### i. Transcript abundances with __[Salmon](https://github.com/COMBINE-lab/salmon)__

We will use a public docker image to run salmon

```{bash salmon}
# run the docker
docker run --rm -it -v $(pwd):/home/my_data combinelab/salmon
ANALYSIS_FOLDER="my_data/Datasets_analysis/TCGA_BRCA"

# run the samples

for fn in "${ANALYSIS_FOLDER}"/downloaded_samples/*.fastq.gz;   
do  samp=`basename ${fn}`;   
regex="${samp%%.fastq.gz}";   
echo "Processing sample ${samp} start: $(date)";   
salmon quant -i my_data/human_data/indexes/GRCh38_v34_public_salmon  \
-l A -r ${fn} --seqBias --gcBias --numBootstraps 100  -p 12 \
--validateMappings --writeMappings="${ANALYSIS_FOLDER}/quants/${regex}.sam" \
-o "${ANALYSIS_FOLDER}/quants/${regex}_quant"; 
echo "end:$(date)";
done
exit

docker run --rm -ti -v $(pwd):/home/my_data  congelos/sncrna_workflow
#save as bam files
for file in "${ANALYSIS_FOLDER}"/quants/*.sam;
do 
regex="${file%%.sam}";
echo "Processing sample ${regex} start: $(date)"; 
echo samtools view -O bam -o ${regex}.bam -@ 8 ${file};
echo "end:$(date)";
done
# remove all .sam files
rm ${ANALYSIS_FOLDER}/quants/*.sam
```
### ii. Alignment with STAR
We use the __[STAR](https://github.com/alexdobin/STAR)__ aligner and then
perform quantification with featureCounts of __[Rsubread](https://www.bioconductor.org/packages/release/bioc/html/Rsubread.html)__ package.

```{bash STAR}
for file in  "${ANALYSIS_FOLDER}"/downloaded_samples/*.fastq.gz; 
do 
samp=`basename ${file}`; 
regex="${samp%%.fastq.gz}"; 
echo "Processing sample ${samp} start: $(date)"; 
STAR --genomeDir my_data/human_data/indexes/GRCh38_v34_public_STAR \
--genomeLoad LoadAndKeep --readFilesIn ${file} --readFilesCommand zcat \
--runThreadN 6 --alignIntronMax 1 --outSAMattributes NH HI NM MD \
--outFilterMultimapNmax 100 --outSAMtype BAM SortedByCoordinate \
--limitBAMsortRAM 40000000000 --outReadsUnmapped Fastx \
--outFilterMismatchNmax 1 --outFilterMatchNmin 14 \
--outFileNamePrefix "${ANALYSIS_FOLDER}/star/${regex}_align/${regex}_";  
echo "end:$(date)";
done

exit
```

Next, we run a docker image which includes varius R packages that
will be used furthermore in the downstream analysis following featureCounts
for the exploratory data analysis of piRNA data

### R docker
```{bash docker for R}
docker run --rm -v $(pwd):/home/0 -p 8787:8787 -e PASSWORD=12345 -e USER=$UID congelos/rocker_tidyverse_plus_de_pckages
```

From here on we work on R using a browser, 
we input http://localhost:8787/ on browser, 0 for username and 12345 for password.

### iv. __[featureCounts](http://subread.sourceforge.net/)__
```{r featureCounts}
library(Rsubread)
library(tidyverse)

ANALYSIS_FOLDER <- file.path("Datasets_analysis", "TCGA_BRCA")
  
list.BAM <- list.files(path = file.path(ANALYSIS_FOLDER, "star"), 
                       pattern = ".bam$", 
                       recursive = TRUE, 
                       full.names = TRUE)

path_gtf <- file.path("human_data","sncRNA_piRNBnk_RNACent_GRCh38_v34.gtf")
todate <- format(Sys.time(), "%d_%b_%Y")

fc <- featureCounts(files = list.BAM,
                    annot.ext =  path_gtf,
                    isGTFAnnotationFile = TRUE,
                    GTF.featureType = "exon",
                    GTF.attrType.extra = c("gene_type", "sRNA_id", "seq_RNA"),
                    nthreads = 6,
                    useMetaFeatures = TRUE,
                    allowMultiOverlap = TRUE,
                    minOverlap = 10,
                    largestOverlap = TRUE,
                    fraction = TRUE,
                    strandSpecific = 1,
                    verbose = TRUE,
                    reportReads = "BAM",
                    reportReadsPath = file.path(ANALYSIS_FOLDER, "star")) 

fc %>% 
  write_rds(file = str_glue(file.path(ANALYSIS_FOLDER, "feature_counts_"), 
                            "{todate}.rds"))
```

Next we will follow the instructions in STEP_3_Exploratory_Data_Analysis.Rmd 

## R Session Info
```{r session Info}
sessionInfo()
```
R Under development (unstable) (2021-01-07 r79806)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.1 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.5.0       stringr_1.4.0       dplyr_1.0.2        
 [4] purrr_0.3.4         readr_1.4.0         tidyr_1.1.2        
 [7] tibble_3.0.4        ggplot2_3.3.3       tidyverse_1.3.0    
[10] Rsubread_2.5.3      BiocManager_1.30.10

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        cellranger_1.1.0  pillar_1.4.7      compiler_4.1.0   
 [5] dbplyr_2.0.0      tools_4.1.0       lubridate_1.7.9.2 jsonlite_1.7.2   
 [9] lifecycle_0.2.0   gtable_0.3.0      lattice_0.20-41   pkgconfig_2.0.3  
[13] rlang_0.4.10      reprex_0.3.0      Matrix_1.3-2      cli_2.2.0        
[17] DBI_1.1.0         rstudioapi_0.13   xfun_0.20         haven_2.3.1      
[21] knitr_1.30        withr_2.3.0       xml2_1.3.2        httr_1.4.2       
[25] fs_1.5.0          generics_0.1.0    vctrs_0.3.6       hms_0.5.3        
[29] grid_4.1.0        tidyselect_1.1.0  glue_1.4.2        R6_2.5.0         
[33] fansi_0.4.1       readxl_1.3.1      modelr_0.1.8      magrittr_2.0.1   
[37] backports_1.2.1   scales_1.1.1      ellipsis_0.3.1    rvest_0.3.6      
[41] assertthat_0.2.1  colorspace_2.0-0  stringi_1.5.3     munsell_0.5.0    
[45] broom_0.7.3       crayon_1.3.4  

## We have worked on : 
```{bash versions, eval = FALSE}
[root@localhost GSE124507_brain_project]# cat /etc/*-release

CentOS Linux release 7.8.2003 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"

[root@localhost GSE124507_brain_project]# docker version

Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:27:04 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:25:42 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

[root@localhost GSE124507_brain_project]# git version
git version 1.8.3.1

[root@localhost GSE124507_brain_project]# pigz --version
pigz 2.3.4
```
